File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/18．单位对单位暂存数值.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local getServerTime = ____require_result_0.getServerTime
local GetHandleId = jass.GetHandleId
local function _____53D6_5355_4F4DID(unit)
    if unit == nil or unit == 0 then
        return 0
    end
    return GetHandleId(unit) or 0
end
local function _____53D6_952E(source, target)
    local sourceId = _____53D6_5355_4F4DID(source)
    local targetId = _____53D6_5355_4F4DID(target)
    if sourceId == 0 or targetId == 0 then
        return ""
    end
    return (tostring(sourceId) .. ":") .. tostring(targetId)
end
____exports["创建单位对单位暂存数值"] = function(_____540D_79F0)
    local _____8868 = {}
    local function _____8BFB_53D6_8BB0_5F55(source, target)
        local key = _____53D6_952E(source, target)
        if key == "" then
            return nil
        end
        local record = _____8868[key]
        if record == nil then
            return nil
        end
        if record["到期"] >= getServerTime() then
            return record
        end
        __TS__Delete(_____8868, key)
        return nil
    end
    return {
        ["名称"] = _____540D_79F0,
        ["写入"] = function(source, target, _____503C, _____6301_7EED_79D2)
            local key = _____53D6_952E(source, target)
            local duration = _____6301_7EED_79D2 or 2
            if key == "" or not (duration > 0) then
                return
            end
            _____8868[key] = {
                ["值"] = _____503C,
                ["到期"] = getServerTime() + duration * 1000
            }
        end,
        ["读取"] = function(source, target)
            local ____opt_1 = _____8BFB_53D6_8BB0_5F55(source, target)
            return ____opt_1 and ____opt_1["值"]
        end,
        ["消耗"] = function(source, target)
            local key = _____53D6_952E(source, target)
            if key == "" then
                return nil
            end
            local record = _____8BFB_53D6_8BB0_5F55(source, target)
            if record == nil then
                return nil
            end
            __TS__Delete(_____8868, key)
            return record["值"]
        end,
        ["清空"] = function()
            _____8868 = {}
        end
    }
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/18．单位对单位暂存数值.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;
const { getServerTime } = require("系统.00．核心系统.05．中心计时器") as {
  getServerTime: (this: void) => number;
};

const GetHandleId = jass.GetHandleId as (handle: any) => number;

export interface 单位对单位暂存数值控制器 {
  readonly 名称: string;
  写入(this: void, source: any, target: any, 值: number, 持续秒?: number): void;
  读取(this: void, source: any, target: any): number | undefined;
  消耗(this: void, source: any, target: any): number | undefined;
  清空(this: void): void;
}

interface 单位对单位暂存数值记录 {
  值: number;
  到期: number;
}

function 取单位ID(this: void, unit: any): number {
  if (unit == null || unit === 0) return 0;
  return GetHandleId(unit) || 0;
}

function 取键(this: void, source: any, target: any): string {
  const sourceId = 取单位ID(source);
  const targetId = 取单位ID(target);
  if (sourceId === 0 || targetId === 0) return "";
  return sourceId + ":" + targetId;
}

export function 创建单位对单位暂存数值(this: void, 名称: string): 单位对单位暂存数值控制器 {
  let 表: Record<string, 单位对单位暂存数值记录 | undefined> = {};

  function 读取记录(this: void, source: any, target: any): 单位对单位暂存数值记录 | undefined {
    const key = 取键(source, target);
    if (key === "") return undefined;
    const record = 表[key];
    if (record == null) return undefined;
    if (record.到期 >= getServerTime()) return record;
    delete 表[key];
    return undefined;
  }

  return {
    名称,
    写入: function 写入(this: void, source: any, target: any, 值: number, 持续秒?: number): void {
      const key = 取键(source, target);
      const duration = 持续秒 ?? 2;
      if (key === "" || !(duration > 0)) return;
      表[key] = { 值, 到期: getServerTime() + duration * 1000 };
    },
    读取: function 读取(this: void, source: any, target: any): number | undefined {
      return 读取记录(source, target)?.值;
    },
    消耗: function 消耗(this: void, source: any, target: any): number | undefined {
      const key = 取键(source, target);
      if (key === "") return undefined;
      const record = 读取记录(source, target);
      if (record == null) return undefined;
      delete 表[key];
      return record.值;
    },
    清空: function 清空(this: void): void {
      表 = {};
    },
  };
}

export {};
